### ISSUE TITLE
Key::TAB / postValue vs. Chrome 53

### ISSUE THREAD

**Issue by lifeofguenter** [minkphp/MinkSelenium2Driver#243]
It seems that with recent Chrome (v53) `Key::TAB` as used in here: https://github.com/minkphp/MinkSelenium2Driver/blob/master/src/Selenium2Driver.php#L681 will actually cause the character `\t` to be added into the input-box, instead of "hitting the tab key".

I can not reproduce it on a normal workstation setup (e.g. real desktop), but it does not work on a headless chrome setup (Ubuntu 15.10 / selenium-server-standalone-2.53.1.jar):

```
$ google-chrome-stable --version
Google Chrome 53.0.2785.92
```

```
$ chromedriver --version
ChromeDriver 2.23.409687 (c46e862757edc04c06b1bd88724d15a5807b84d1)
```

Can be reproduced (headless-browser) with the following snippet:

```
<?php

ini_set('display_errors', true);
error_reporting(E_ALL);

require 'vendor/autoload.php';

$driver = new \Behat\Mink\Driver\Selenium2Driver('chrome');
$session = new \Behat\Mink\Session($driver);

$session->start();

$session->visit('http://localhost/admin/login');
sleep(2);
$session->getPage()->fillField('email', 'test@test.test');
$session->getPage()->fillField('password', 'abcdef');
$session->getPage()->pressButton('login');
```

selenium-server is started like this:

```
screen -S selenium -d -m sh -c "DBUS_SESSION_BUS_ADDRESS=/dev/null xvfb-run -s '-screen 0 1024x768x24' -a java -jar /opt/selenium-server/selenium-server-standalone.jar > /home/bamboo/selenium.stdout.log 2> /home/bamboo/selenium.stderr.log"
```

```
$ java -version
java version "1.8.0_77"
Java(TM) SE Runtime Environment (build 1.8.0_77-b03)
Java HotSpot(TM) 64-Bit Server VM (build 25.77-b03, mixed mode)
```

Swapping out the browser from chrome to firefox (version 46.0.1) works, as well as removing that `Key::TAB` from the code above (however not sure if it is the right fix, as obviously it should be working?).

**Comment by aik099** [comment]
1. maybe this is bug in Selenium, that changes in Chrome (in way how TAB is handled) are not handled properly
2. the TAB is sent to trigger `onchange` JS event

**Comment by lifeofguenter** [comment]
Is this maybe related to: https://github.com/instaclick/php-webdriver/issues/70 ?

**Comment by adridev** [comment]
Facing the same issue here.

Google Chrome 53.0.2785.92 

ChromeDriver 2.23.409687 (c46e862757edc04c06b1bd88724d15a5807b84d1)

Selenium server 2.53.1

**Comment by fonsecas72** [comment]
Did anyone found or confirmed anything?

**Comment by aik099** [comment]
Me nope. Probably somebody needs to post an issue on Chrome Driver issue tracker (with a link to it posted back here).

**Comment by fonsecas72** [comment]
ok, just for others to see, if I'm not mistaken, a workaround for `findField` could be something like this:

```php
$xpath = $this->getSession()->getSelectorsHandler()->selectorToXpath('named', [
    'field',
    $locator
]);
$elementName = strtolower($this->getSession()->getDriver()->getTagName($xpath));
if ($elementName === 'input' || $elementName === 'textarea') {
    $element = $this->getSession()->getDriver()->getWebDriverSession()->element('xpath', $xpath);
    $element->postValue(['value' => [$value]]);
} else {
    $this->findField($locator)->setValue($value); 
}
```

**Comment by aik099** [comment]
> https://github.com/minkphp/MinkSelenium2Driver/issues/243#issuecomment-245890248

@fonsecas72 , you've just removed TAB. That is one of possible solutions, but then `onchange` event won't be fired in JavaScript when value is changed.

**Comment by fonsecas72** [comment]
yup, users should be aware of the downside when using it. I'd suggest to move it to a special method or something :)

@aik099 if you have a cleaner solution please let me know.

**Comment by adridev** [comment]
for me, the workaround was to remove the Key::TAB from the end of the line in Selenium2Driver.php:662.

About the "change" event, I don't know if is correct to fire this event automatically when an input is filled. I think could be more natural to have to explicitly hit the enter or leave the field to fire this event.

**Comment by aik099** [comment]
That `change` event firing also created problems for testing stuff, like jQueryUI.Autocomplete, because immediately after value change the dropdown with completions was hidden.

@stof , I know this might be a BC break, but maybe for inputs with free text entry (not checkbox/radio button/select) we can stop firing `change` JS event. If agreed, then we need a way to stop firing `change` event for other drivers as well.

**Comment by lifeofguenter** [comment]
I think first we need to figure out if this is a BC break intended by chromedriver/chrome, or if it is actually a bug.

**Comment by adorin** [comment]
How about this? https://github.com/adorin/MinkSelenium2Driver/commit/e2710a2d3fd2df46182bf0482d7862630fae7a8c

**Comment by adridev** [comment]
I'm making the same as a workaround but I not sure if is there the correct place for calling to excuteScript. 

Probably is more natural to call the blur method after calling to setValue within the step definition. 

What do you think guys?

**Comment by adorin** [comment]
The idea is to fix the issue without otherwise altering functionality.

I wouldn't combine a bugfix with a change like that.

**Comment by adridev** [comment]
It's true, you are right.

**Comment by aik099** [comment]
> How about this? https://github.com/adorin/MinkSelenium2Driver/commit/e2710a2d3fd2df46182bf0482d7862630fae7a8c

@adorin , the `$this->executeScript("document.activeElement && document.activeElement.blur()");` can be replaced with `$this->blur($xpath);`. As long as test pass that would solve this problem.

@adridev , I've suggested an update to @adorin solution according to your suggestion.

**Comment by kwoxer** [comment]
Same for me in Chrome Version 53.0.2785.101.

It helped to delete Key::TAB in line 681 

> $value = str_repeat(Key::BACKSPACE . Key::DELETE, $existingValueLength) . $value . Key::TAB;

**Comment by kwoxer** [comment]
Also https://github.com/minkphp/MinkSelenium2Driver/issues/194 seems to be identical.

**Comment by cocciagialla** [comment]
@lifeofguenter it seems that the TAB doesn't work only for password type inputs. 

For example, this worked for me:
`
// Add the TAB key to ensure we unfocus the field as browsers are triggering the change event only
// after leaving the field.
$value = str_repeat(Key::BACKSPACE . Key::DELETE, $existingValueLength) . $value;
if ($elementType != 'password'){
   $value .= Key::TAB;
}
`
I think is not a BC break to exclude only password type.

**Comment by aik099** [comment]
@cocciagialla , did #244 work for you as well? A bonus side effect from #244 is ability to test jQueryUI.Autocomplete and similar controls.

**Comment by cocciagialla** [comment]
@aik099 yes it works perfectly. I hope it will be merged soon.

**Comment by aik099** [comment]
Just merged #244. Happy test running.
